### ISSUE TITLE
[clang-tidy] False positive `cppcoreguidelines-pro-type-static-cast-downcast` when base type is not polymorphic

### ISSUE THREAD

**Issue by ChrisThrasher** [llvm/llvm-project#69414]
https://godbolt.org/z/ddE4bYnse

```cpp
struct Base
{
    int i;
};

struct Derived : Base
{
    float f;
};

int main()
{
    Derived d1;
    Base& base = d1;
    Derived& d2 = static_cast<Derived&>(base);
}
```

`cppcoreguidelines-pro-type-static-cast-downcast` emits a warning even when the base class is not polymorphic. In the above case the `static_cast` is necessary to cast from base to derived since a `dynamic_cast` is not allowed.

**Comment by PiotrZSL** [comment]
Then make that class polymorphic.

**Comment by ChrisThrasher** [comment]
It's not a class I control. It's all 3rd party types.

**Comment by PiotrZSL** [comment]
I could add a "StrictMode" option to a check, with 'true' would check all casts, with false would check only on polymorphic types. Could be ?

**Comment by ChrisThrasher** [comment]
My thought was that the check would detect non-polymorphic base classes and omit the diagnostic in such cases. Is that feasible?

**Comment by PiotrZSL** [comment]
Yes, but I would put this under config option.

**Comment by ChrisThrasher** [comment]
What's the use case for the example I shared above? If the base class is not polymorphic how else are you supposed to case from base to derived? That doesn't seem like a useful diagnostic unless there is an alternative way of doing such a thing.

**Comment by PiotrZSL** [comment]
Simply you don't, proper solution would be to design code so cast wouldn't be needed.
